src/lfortran/semantics/ast_body_visitor.cpp

```diff
@@ -2117,6 +2117,23 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         head.m_end = end;
         head.m_increment = increment;
         if (head.m_v != nullptr) {
+            // iterate over body and for every assignment check target, if it is equal to head.m_v, throw error
+            for (size_t i = 0; i < body.size(); i++) {
+                ASR::stmt_t* stmt = body[i];
+                if (stmt->type == ASR::stmtType::Assignment) {
+                    ASR::Assignment_t* assign = ASR::down_cast<ASR::Assignment_t>(stmt);
+                    ASR::expr_t* target = assign->m_target;
+                    if (target->type == ASR::exprType::Var) {
+                        ASR::Var_t* var = ASR::down_cast<ASR::Var_t>(target);
+                        if (var->m_v->type == ASR::symbolType::Variable) {
+                            ASR::Variable_t* target_var = ASR::down_cast<ASR::Variable_t>(var->m_v);
+                            if (to_lower(x.m_var) == std::string(target_var->m_name)) {
+                                throw SemanticError("Do loop: assignment to loop variable `" + std::string(target_var->m_name) +"` is not allowed", assign->base.base.loc);
```

Comment:
I would highlight just the target variable:
```suggestion
                                throw SemanticError("Assignment to loop variable `" + std::string(target_var->m_name) +"` is not allowed in a do loop", target_var->base.base.loc);
```